### ISSUE TITLE
Native type on multiple argument custom type

### ISSUE THREAD

**Issue by plicease** [Perl5-FFI/FFI-Platypus#18]
Basically a series of multiple arguments get the same ffi type right now.  It doesn't matter as far as the arguments API, but it means that `libffi` might be using the wrong type.  This probably is okay for integers on intel, but may barf on big endian or floats.

**Comment by plicease** [comment]
This does actually require a (compatible) interface change.  Something like this:

```perl
sub ffi_custom_type_api_1
{
  {
    native_type         => 'opaque',
    perl_to_native      => \&perl_to_native,
    perl_to_native_post => \&perl_to_native_post,
    argument_count      => 2,
  }
}
```

Should be something more like this:

```perl
sub ffi_custom_type_api_1
{
  {
    native_type         => [ 'opaque', 'size_t' ],
    perl_to_native      => \&perl_to_native,
    perl_to_native_post => \&perl_to_native_post,
    argument_count      => 2,
  }
}
```

**Comment by pipcet** [comment]
I've been experimenting on my fork here: https://github.com/pipcet/FFI-Platypus/tree/recursive-types and, uh, got a little carried away. I kept the API compatible but restructured much of the calling code. Among other things, this issue would be fixed, though, using the API you suggested.

However, I must say it's very difficult to test whether it's actually working, because I'm running a 64-bit system and all our types appear to be passed as 64 bits. gdb appears to be indicating it's doing the right thing, though.

Since the diff is quite large, I can't really in good conscience ask you to go through it and let me know what, if anything, you might consider for inclusion. The main change is that I broke up the code in ffi_platypus_call.h into separate functions, even though that means we can no longer use alloca() (as an idea, we could use runtime-sized arrays instead, but I understand that's a difficult subject with Windows compilers); apart from being more maintainable, that allows us to call the argument preparation/mangling functions recursively, so a ->custom_type can return another ->custom_type and everything appears to be working (https://github.com/pipcet/FFI-Platypus/blob/recursive-types/t/ffi_platypus_type_rectangle_5.t might serve as an example).

The main change as far as the actual design goes is to reference types by their SV *s internally, and letting Perl handle their refcounts that way; I've also converted types to be hashref objects so more of the code can be written in Perl rather than C.

Sorry for going a bit off-topic.

**Comment by plicease** [comment]
@pipcet It looks like an interesting (and exciting) experiment, and I am highly envious of your `ffi_platypus_call.h`, unfortunately this approach is much slower than the current implementation.  I use FFI-Performance (in my github but not on CPAN) to compare FFI implementations for simple use cases (I don't compare things like custom types since Platypus is the only one to support such a feature):

Platypus 0.31:
```
## 01 single integer argument and return ##

Benchmark: timing 10485760 iterations of Inline, OO, Platypus, PlatypusSl, Raw, SweetLite, SweetLite2, XS...
    Inline:  1 wallclock secs ( 1.52 usr +  0.00 sys =  1.52 CPU) @ 6898526.32/s (n=10485760)
        OO:  5 wallclock secs ( 4.49 usr +  0.00 sys =  4.49 CPU) @ 2335358.57/s (n=10485760)
  Platypus:  4 wallclock secs ( 2.97 usr +  0.00 sys =  2.97 CPU) @ 3530558.92/s (n=10485760)
PlatypusSl:  8 wallclock secs ( 7.29 usr +  0.00 sys =  7.29 CPU) @ 1438375.86/s (n=10485760)
       Raw:  8 wallclock secs ( 7.79 usr +  0.00 sys =  7.79 CPU) @ 1346053.92/s (n=10485760)
 SweetLite: 13 wallclock secs (13.21 usr +  0.00 sys = 13.21 CPU) @ 793774.41/s (n=10485760)
SweetLite2:  3 wallclock secs ( 2.93 usr +  0.00 sys =  2.93 CPU) @ 3578757.68/s (n=10485760)
        XS:  0 wallclock secs ( 1.35 usr +  0.00 sys =  1.35 CPU) @ 7767229.63/s (n=10485760)

                Rate SweetLite  Raw PlatypusSl   OO Platypus SweetLite2 Inline   XS
SweetLite   793774/s        -- -41%       -45% -66%     -78%       -78%   -88% -90%
Raw        1346054/s       70%   --        -6% -42%     -62%       -62%   -80% -83%
PlatypusSl 1438376/s       81%   7%         -- -38%     -59%       -60%   -79% -81%
OO         2335359/s      194%  73%        62%   --     -34%       -35%   -66% -70%
Platypus   3530559/s      345% 162%       145%  51%       --        -1%   -49% -55%
SweetLite2 3578758/s      351% 166%       149%  53%       1%         --   -48% -54%
Inline     6898526/s      769% 412%       380% 195%      95%        93%     -- -11%
XS         7767230/s      879% 477%       440% 233%     120%       117%    13%   --
```

recursive types:
```
## 01 single integer argument and return ##

Benchmark: timing 10485760 iterations of Inline, OO, Platypus, PlatypusSl, Raw, SweetLite, SweetLite2, XS...
    Inline:  1 wallclock secs ( 1.45 usr +  0.00 sys =  1.45 CPU) @ 7231558.62/s (n=10485760)
        OO:  3 wallclock secs ( 4.54 usr +  0.00 sys =  4.54 CPU) @ 2309638.77/s (n=10485760)
  Platypus: 62 wallclock secs (62.91 usr +  0.00 sys = 62.91 CPU) @ 166678.75/s (n=10485760)
PlatypusSl: 78 wallclock secs (77.49 usr +  0.01 sys = 77.50 CPU) @ 135300.13/s (n=10485760)
       Raw:  9 wallclock secs ( 9.14 usr +  0.00 sys =  9.14 CPU) @ 1147238.51/s (n=10485760)
 SweetLite: 17 wallclock secs (17.47 usr +  0.00 sys = 17.47 CPU) @ 600215.23/s (n=10485760)
SweetLite2: 71 wallclock secs (71.16 usr +  0.04 sys = 71.20 CPU) @ 147271.91/s (n=10485760)
        XS:  1 wallclock secs ( 1.64 usr +  0.00 sys =  1.64 CPU) @ 6393756.10/s (n=10485760)

                Rate PlatypusSl SweetLite2 Platypus SweetLite  Raw   OO   XS Inline
PlatypusSl  135300/s         --        -8%     -19%      -77% -88% -94% -98%   -98%
SweetLite2  147272/s         9%         --     -12%      -75% -87% -94% -98%   -98%
Platypus    166679/s        23%        13%       --      -72% -85% -93% -97%   -98%
SweetLite   600215/s       344%       308%     260%        -- -48% -74% -91%   -92%
Raw        1147239/s       748%       679%     588%       91%   -- -50% -82%   -84%
OO         2309639/s      1607%      1468%    1286%      285% 101%   -- -64%   -68%
XS         6393756/s      4626%      4241%    3736%      965% 457% 177%   --   -12%
Inline     7231559/s      5245%      4810%    4239%     1105% 530% 213%  13%     --
```

(PlatypusSl, and SweetLite2 are FFI::Platypus based implementations).

I'd really like to see this sort of flexibility in Platypus, and I don't mind custom types itself being slow (if you need that degree of flexibility there will be a cost since you have to dip back into Perl space and back anyway), but not at the expense of the really simple stuff being slow.  I might borrow some of your code it looks like you did a lot of good work, or maybe if you can rework it into something that works with the current arch a little better.  Another thing that I would like to do is abstract some of the implementation details to make alternate implementations easier (see #44).  That would allow us to work on experimental stuff like this in parallel.

One minor thing is that we can't use `//` in Platypus because we still support Perl 5.8.1 so long as its supported by `Module::Build`.  I am hoping that won't be too much longer but it is hard to say.

**Comment by pipcet** [comment]
Thanks for your comments, that's really encouraging! I've tracked down the performance issue to the use of sv_derived_from in an inner loop, which is obviously a bug I shall fix. I hadn't known about FFI::Performance, it's much appreciated that you pointed it out to me. (In fact it seems one of the tests doesn't run at all! I'll have to look into that.)

I'm aware of some other performance issues still, and I absolutely agree we should not merge anything that hurts performance particularly of the integer code. That said, an (unfair) comparison between my code slightly hacked to avoid malloc appears to be indistinguishable in performance from FFI::Platypus 0.31.

Thanks for pointing out my mistake in using //, that was unintentional.

Do feel free to take any of my code, obviously, but also feel free to ask me for pull requests for those features you want. I'm more than happy to split up things and polish them for inclusion.

Thanks again,
Philip

**Comment by plicease** [comment]
@pipcet I have been thinking about this and I am not entirely sold on using Perl types to represent types.  One of the goals of this project is to support alternate versions of Perl that do not have XS.  I guess it depends on just how just how much can be done in Perl vs C, and how much overhead is involved.

**Comment by pipcet** [comment]
I think the right way is to use method calls, whether or not there are separate classes for underlying types. In recursive-types, all libffi types have class 'FFI::Platypus::Type::FFI', and the actual methods contain switch statements which use C enumerations, not sv_derived_from or anything that depends on Perl's type system.

For the default implementation I'd defend my decision to use hashrefs and deep classes that they're blessed into, because that way you don't need to maintain a central header file with a huge enum that keeps growing as types are added, merged, and replaced. For implementations that have such an enum anyway, such as a pure FFI pass-through version without complex types, simply use one class and store the enum value in the pointer.

Or was what you meant using SV *s rather than C pointers in our implementation? Again, that's an implementation detail, but I'd really like to keep it that way because it gives us refcounting and anonymous/renamed types, which are nice to have in theory and might be useful in practice when interfacing with closure-rich environments.

I think all of this hinges on getting started on #44, because it's hard to tell what's a design decision and what's an implementation detail as long as there's only a single implementation.

(As for the performance issues, I've committed some fixes and performance is now within 10% of Platypus 0.31. It's hard to be sure given the variance, but I think the actual speed difference, on my machine, is closer to zero than that, and it's not consistently going one way or the other.)

**Comment by plicease** [comment]
@pipcet Not committing to anything yet, just wanted make this goal clear.  I don't see ref counting as a big plus for the reasons that I mentioned earlier, but overall I think your trade offs are defensible.

**Comment by pipcet** [comment]
I've been thinking about it, and running your new benchmarks, and it now seems to me Perl hashes are just too slow for our (relatively) lean code (hashing the key in advance seems to make no difference). So my current idea is to store a void * for the getter routines to use along with the SV * that references the hash, and hopefully that will make performance of the simple benchmarks comparable again. I guess that's the point of experiments, to figure out what doesn't work.

My main reason for keeping the SV * at all is to actually encourage types to be written in Perl rather than C except for the really performance-sensitive stuff, and to think of the void * only as a cached hash entry. (The "obvious" way of writing a type in perl is to return a closure cast to opaque from the native_to_perl_pointer etc. methods, but by explicitly passing the SV * we can avoid a circular reference).

So, such hybrid objects would be my current proposal. I'll see about putting my code where my mouth is soon unless you object.

Thank you again for your insightful comments, and the wonderful work you've done with Platypus so far.

**Comment by plicease** [comment]
Using a closure for custom types would also make it easier to write a custom type C, which is a long term goal.

**Comment by plicease** [comment]
I am changing this to a feature request from a bug and closing.  It's never been supported in practice so it isn't really a bug.  The documentation should probably be updated though.  There are also other ways to handle this.  The advantage of doing it in XS is that it would be a bit more performant.  The disadvantage is that it overly complicates the innards.  I think atm at least and given the work load the cons outweigh the benefits.
